src/layer/place.js

```diff
@@ -187,7 +181,124 @@ export const city = {
         [11, 0.9],
       ],
     },
-    "text-field": Label.localizedName,
+    "text-field": [
+      "let",
+      "localizedName",
+      "",
+      "localizedCollator",
+      ["collator", {}],
+      "diacriticInsensitiveCollator",
+      ["collator", {}],
+      [
+        "case",
+        [
+          "==",
+          ["var", "localizedName"],
+          ["get", "name"],
+          ["var", "localizedCollator"],
+        ],
+        ["var", "localizedName"],
+        // If the name in the preferred language is the same as the name in the local language except for the omission of diacritics and/or the addition of a suffix (e.g., "City" in English), replace the common prefix with the local name.
+        [
+          "all",
+          [
+            "==",
+            ["slice", ["var", "localizedName"], 0, ["length", ["get", "name"]]],
+            ["get", "name"],
+            ["var", "diacriticInsensitiveCollator"],
+          ],
+          [
+            "in",
+            [
+              "slice",
+              // "Montreal" vs. "Montréal"
+              ["concat", ["var", "localizedName"], " "],
+              ["length", ["get", "name"]],
+              ["+", ["length", ["get", "name"]], 1],
+            ],
+            // "Quebec City" vs. "Québec", "Washington, D.C." vs. "Washington"
+            " ,",
+          ],
+        ],
+        [
+          "concat",
+          ["get", "name"],
+          ["slice", ["var", "localizedName"], ["length", ["get", "name"]]],
+        ],
+        // If the name in the preferred language is the same as the name in the local language except for the omission of diacritics and/or the addition of a prefix (e.g., "City of" in English or "Ciudad de" in Spanish), replace the common suffix with the local name.
+        [
+          "all",
+          [
+            "==",
+            [
+              "slice",
+              ["var", "localizedName"],
+              [
+                "-",
+                ["length", ["var", "localizedName"]],
+                ["length", ["get", "name"]],
+              ],
+            ],
+            ["get", "name"],
+            ["var", "diacriticInsensitiveCollator"],
+          ],
+          [
+            "==",
+            [
+              "slice",
+              ["var", "localizedName"],
+              [
+                "-",
+                [
+                  "-",
+                  ["length", ["var", "localizedName"]],
+                  ["length", ["get", "name"]],
+                ],
+                1,
+              ],
+              [
+                "-",
+                ["length", ["var", "localizedName"]],
+                ["length", ["get", "name"]],
+              ],
+            ],
+            " ",
+          ],
```

Comment:
This would be just a simple `in` expression, except that we still need to do diacritic folding first. Unfortunately, there’s no expression for returning a string with diacritics removed, and the `in` expression doesn’t let us pass in a collator either: mapbox/mapbox-gl-js#9339.